HTML block: fix parsing

ellatrix opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description

Fixes #24282, or at least partly.
The issue has been around since the beginning of Gutenberg.

The problem is that, when parsing the blocks, `element.innerHTML` encodes these characters.

What we're looking for is the raw block content, so there's no need to parse the HTML at all, it can be skipped.

## How has this been tested?

Type `0 < 1` in an HTML block. Save and reload. There should be no error.

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ellatrix commented:
Asking wider feedback because it adds a small piece to an existing API.

youknowriad commented on packages/block-library/src/html/block.json:
```diff
@@ -5,7 +5,7 @@
 		"content": {
 			"type": "string",
-			"source": "html"
+			"source": "raw"
```
I'm pretty sure there's already a PR that does the same thing with lengthy discussion from  @aduth

ellatrix replied:
Oh haha :)

youknowriad replied:
I didn't find it :(

ellatrix replied:
Found it #25120. I was actually just talking with @getdave about this to try to find a solution.

youknowriad replied:
I found it, this is the one I was talking about https://github.com/WordPress/gutenberg/pull/10551

ellatrix replied:
Hm, this doesn't have any comments from @aduth though.

I'm also not sure that PR is a good solution, because it will break blocks that may use innerHTML differently.

ellatrix replied:
@youknowriad Nice, thanks, will read quick.

ellatrix replied:
So the difference with that PR is that it still uses the `html` type instead of a new `raw` type, but the solution is essentially the same.

youknowriad replied:
yes, and I believe @aduth's point is that it's the `html` type itself that is "entirely" broken

getdave replied:
Although...if this fixes the immediate experience for users could we land this PR and then write a followup Issue detailing the problems with `html` and how we might look to resolve them long term? Not being able to put HTML in the HTML block feels like a bit of a poor UX 😄

youknowriad replied:
I'm not sure we should add a new API (new "source" type)

getdave reviewed (approved):
I tested this using [the examples provided by the contributor in the original Issue](https://github.com/WordPress/gutenberg/issues/24282#issuecomment-702900355).

This PR does seem to resolve the issue in that chars such as `<` are not converted to entities. Here's the comparison from my testing.

| Master      | This PR  | 
| :------------- | :----------: | 
|  <img width="1280" alt="Screenshot 2020-11-27 at 09 12 16" src="https://user-images.githubusercontent.com/444434/100432013-e5936c80-3090-11eb-9a80-0d828c6ff7bd.png"> | <img width="1280" alt="Screen Shot 2020-11-27 at 09 11 42" src="https://user-images.githubusercontent.com/444434/100432019-e926f380-3090-11eb-97e4-4ea0dda85e19.png">   | 

As you can see on master we get errors because the generated save content doesn't match the post content. On this PR however, there are no errors and everything is great 🎉 

@ellatrix I'll defer to yourself and @youknowriad regarding whether to land this or to try and ship a more comprehensive fix to HTML in general. 

One thing - we should probably add some tests to cover this scenario.

ellatrix commented:
Definitely needs some tests. This is just a proposal so far. I'd like some feedback from at least some more people like @youknowriad @mtias @mcsf.

I don't see a good alternative here. We could try to fix the `html` sourcing (which relies on `Element.innerHTML` right now), but that's not easy to do (I think) and it would be a breaking change on an existing API vs an addition to the API. The simplest way is to allow a block to access the raw block content.

getdave commented:
> We could try to fix the html sourcing (which relies on Element.innerHTML right now)

I believe this is why the entities are encoded. See this note from [innerHTML from MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML):

```
Note: If a <div>, <span>, or <noembed> node has a child text node that 
includes the characters (&), (<), or (>), innerHTML returns these characters 
as the HTML entities "&amp;", "&lt;" and "&gt;" respectively. Use 
Node.textContent to get a raw copy of these text nodes' contents.
```

ellatrix commented:
Yes, because of this, the HTML you SET with `Element.innerHTML` may not be equal to the HTML you GET with `Element.innerHTML`.

One alternative solution here is to always encode characters just like `innerHTML` does. I think it's just the opening bracket that's causing the problem.

getdave commented:
> One alternative solution here is to always encode characters just like innerHTML does.

This _sounds_ similar to [what my original PR did but I was _decoding_](https://github.com/WordPress/gutenberg/pull/25120/files#diff-10068f6ecaecb2a75549e1b4a071cb1a1ef51fc88586ba998b5a66db30c73cc1R46). 

Correct me if I'm wrong but the key would be to ensure that the Gutenberg annotations are 1:1 with whatever the user types into the Block? If the database needs to have that data encoded for persistence then that's ok so long as when it's parsed back it returns back 1:1 with what the user originally entered?

aristath commented:
The parser has changed since this PR was created, do we still want to do this? Or should we close this?

ellatrix commented:
@aristath Unfortunately, this is still a problem.

Would be good to agree on the solution.

getdave committed:
HTML block: fix parsing

getdave committed:
Reinstate the change in the new file

getdave commented:
I've rebased the branch. As @aristath said the parser changed and the original change was in a file that no longer existed.

I've moved to what seems like the new location and reinstated @ellatrix's original change.

getdave commented:
So to recap...

The problem is that when HTML code it put into the HTML block certain characters are encoded into HTML entities.

The reason for this is that the `html` source matcher used by the `core/html` block relies on `Element.innerHTML` [which will encode certain characters](https://github.com/WordPress/gutenberg/pull/27268#issuecomment-735883066).

@ellatrix's proposed fix (here) is to introduce a new `raw` source type which simply returns the block content as is. This avoids/bypasses the `innerHTML` issue and we get a 1:1 with what the user entered.

@youknowriad [was not in favour](https://github.com/WordPress/gutenberg/pull/27268#discussion_r530395618) of introducing a new API however. He referenced [a previous PR](https://github.com/WordPress/gutenberg/pull/10551) by @johngodley where they were looking to address a similar issue by either

- fixing upstream in `hpq` 
- bypassing `hpq` entirely [based on the presence of a `selector` in the attribute schema](https://github.com/WordPress/gutenberg/pull/10551/files#r239020521)

The outcome of https://github.com/WordPress/gutenberg/pull/10551 is that neither @youknowriad nor @aduth were happy to commit to merging the PR. However, the goal of that PR seems to have been around normalizing the HTML whereas our problem is relating to the HTML being fundamentally transformed.

There is also https://github.com/WordPress/gutenberg/pull/25120 which attempts to "fix" the `html` parser at the point of `Element.innerHTML` by converting the HTML entities back into their string form. The description of that PR also includes a lot of detail about why this bug occurs so it's worth reading for context.

dmsnell committed:
Pass un-parsed innerHTML into getAttribute to support raw property

In #39424 when we added an optimization to only parse a block's
innerHTML once we also changed the behavior that the innerHTML
propety represented the raw HTML loaded by the parser. Instead,
what we have since that change is the DOM of the parsed HTML.

In this patch we're adding yet-another parameter to the bag of
arguments in `getBlockAttribute()` so that the new `raw` type
can read and reproduce that original source, e.g. when reading
the string `1 < 0` the parsed value's `innerHTML` will be
`1 &lt; 0` even though the block's raw content was `1 < 0`.

dmsnell reviewed (approved):
This won't impact any other blocks besides the HTML block. We should probably try and keep an eye out in case anyone else wants to use `raw` for another block. As discussed in person, I think there are still peculiarities about what the role of the HTML block even is or what _proper_ behavior here is.

So practically I think this is a fair tactical change and addresses the issue at hand without adding much baggage. If we ever get around to reworking the HTML block we can change this stuff here, or if we overhaul the attribute sourcing and validation.

1 > 0

ellatrix committed:
Add e2e test

getdave commented:
I'm in favour of merging this one. The impact seems limited to the HTML block and it is "opt in" rather than default behaviour.

Current unit test failures seem related and will need to be resolved prior to merge.

ellatrix committed:
Add to schema

getdave commented:
I'm very pleased to see this get merged 🥳
